src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -0,0 +1,588 @@
+
+                    let (block, filter_data, uncompressed_size) = first
+                        .current_sstable()
+                        .download_next_block()
+                        .await?
+                        .unwrap();
+                    let algorithm = Block::get_algorithm(&block)?;
+                    if algorithm == CompressionAlgorithm::None
+                        && algorithm != self.compression_algorithm
+                        && self.executor.task_config.gc_delete_keys
+                    {
+                        first
+                            .current_sstable()
+                            .init_block_iter(block, uncompressed_size)?;
+                        break;
+                    }
+
+                    let (stale_count, total_count) =
+                        first.estimate_key_count(uncompressed_size as u64);
+                    skip_raw_block_size += block.len() as u64;
+                    skip_raw_block_count += 1;
+                    let largest_key = first.current_sstable().current_block_largest();
+
+                    self.executor
+                        .builder
+                        .add_raw_block(
+                            block,
+                            filter_data,
+                            smallest_key,
+                            largest_key,
+                            uncompressed_size,
+                            stale_count,
+                            total_count,
+                        )
+                        .await?;
+                    self.executor.clear();
+                }
+                if !first.current_sstable().is_valid() {
+                    first.next_sstable().await?;
+                    continue;
+                }
+                first.init_block_iter().await?;
+            }
+
+            let target_key = second.current_sstable().key();
+            let iter = first.sstable_iter.as_mut().unwrap().iter.as_mut().unwrap();
+            self.executor.run(iter, target_key).await?;
+            if !iter.is_valid() {
+                first.sstable_iter.as_mut().unwrap().iter.take();
+                if !first.current_sstable().is_valid() {
+                    first.next_sstable().await?;
+                }
+            }
+        }
+        let rest_data = if !self.left.is_valid() {
+            if !self.right.is_valid() {
+                return self.executor.builder.finish().await;
+            }
+            &mut self.right
+        } else {
+            &mut self.left
+        };
+        {
+            let sstable_iter = rest_data.sstable_iter.as_mut().unwrap();
+            let target_key = FullKey::decode(&sstable_iter.sstable.value().meta.largest_key);
+            if let Some(iter) = sstable_iter.iter.as_mut() {
+                self.executor.run(iter, target_key).await?;
+                assert!(!iter.is_valid());
+            }
+            sstable_iter.iter.take();
+        }
+
+        while rest_data.is_valid() {
```

Comment:
Same question: Is there a case where we keep triggering a block direct copy (even at the last level), and thus the tombstone cleanup is not timely enough ?

We need to discuss whether to enable block direct copy at the last level or add more restrictions to avoid affecting tombstone cleanup. cc @hzxa21 @Little-Wallace

Reply:
Good catch. There is a problem that it this PR can not calculate the number of delete-range and tombstone. 
I think we shall keep all version of one key in one block and record tombstone number in Block-Meta.

Reply:
Will @Li0k 's question be resolved in this PR?

Reply:
Yes. I have collect the count of tombstone accurately in this PR. So that if some block owns too many tombstone, meta will trigger a spacial task by `TombstoneReclaimCompactionPicker`.